lib/rack/session/dalli.rb

```diff
@@ -4,75 +4,143 @@
+      # "rack:session".
+      #
+      # Whether you are using the `:cache` option or not, it is not recommend
+      # to set `:expires_in`. Instead, use `:expire_after`, which will control
+      # both the expiration of the client cookie as well as the expiration of
+      # the corresponding entry in memcached.
+      #
+      # Rack::Session::Dalli also accepts a host of options that control how
+      # the sessions and session cookies are managed, including the
+      # aforementioned `:expire_after` option. Please see the documentation for
+      # Rack::Session::Abstract::Persisted for a detailed explanation of these
+      # options and their default values.
+      #
+      # Finally, if your web application is multithreaded, the
+      # Rack::Session::Dalli middleware can become a source of contention. You
+      # can use a connection pool of Dalli clients by passing in the
+      # `:pool_size` and/or `:pool_timeout` options. For example:
+      #
+      #   use Rack::Session::Dalli,
+      #     :memcache_server => "mc.example.com:1234",
+      #     :pool_size => 10
+      #
+      # You must include the `connection_pool` gem in your project if you wish
+      # to use pool support. Please see the documentation for ConnectionPool
+      # for more information about it and its default options (which would only
+      # be applicable if you supplied one of the two options, but not both).
+      #
       def initialize(app, options={})
         super
-        @mutex = Mutex.new
-        mserv = @default_options[:memcache_server]
-        mopts = @default_options.reject{|k,v| !DEFAULT_OPTIONS.include? k }
-        @pool = options[:cache] || ::Dalli::Client.new(mserv, mopts)
-        @pool.alive!
-      end
 
-      if defined?(Abstract::Persisted)
-        def find_session(req, sid)
-          get_session req.env, sid
-        end
+        # determine the default TTL for newly-created sessions
+        @default_ttl = ttl @default_options[:expire_after]
 
-        def write_session(req, sid, session, options)
-          set_session req.env, sid, session, options
-        end
+        mopts = DEFAULT_DALLI_OPTIONS.merge \
+          options.reject {|k, _| DEFAULT_OPTIONS.key? k }
+        mserv = mopts.delete :memcache_server
 
-        def delete_session(req, sid, options)
-          destroy_session req.env, sid, options
-        end
-      end
+        pool_options = {}
+        pool_options[:size] = mopts.delete :pool_size if mopts.key? :pool_size
+        pool_options[:timeout] = mopts.delete :pool_timeout if mopts.key? :pool_timeout
 
-      def generate_sid
-        while true
-          sid = super
-          break sid unless @pool.get(sid)
-        end
+        @pool =
+          if options[:cache] and options[:cache].respond_to? :with
+            options[:cache]
+          elsif !pool_options.empty?
+            mopts[:threadsafe] = false unless mopts.key? :threadsafe
```

Comment:
Ah, got it.  This only applies in the pool case, when the lock is held at the pool level.

Again, if moved to the proposed `extract_dalli_options` method described above, with an appropriate line comment (e.g. 'For a connection pool, locking is handled at the pool level') this would be a lot clearer.